File: apps/mobile-app/ios/VaultStoreKit/Services/VaultMergeService.swift
import Foundation
import SQLite3
import CryptoKit

/// Service for vault merge operations using the Rust core library.
/// Wraps UniFFI-generated bindings for LWW merge operations on SQLite vault databases.
public class VaultMergeService {
    public static let shared = VaultMergeService()

    private init() {}

    /// Error types for vault merge operations.
    public enum VaultMergeError: Error, LocalizedError {
        case jsonError(String)
        case rustError(String)
        case databaseError(String)
        case invalidInput(String)
        case encryptionError(String)

        public var errorDescription: String? {
            switch self {
            case .jsonError(let msg): return "JSON error: \(msg)"
            case .rustError(let msg): return "Rust error: \(msg)"
            case .databaseError(let msg): return "Database error: \(msg)"
            case .invalidInput(let msg): return "Invalid input: \(msg)"
            case .encryptionError(let msg): return "Encryption error: \(msg)"
            }
        }
    }

    /// Get the list of syncable table names from Rust.
    public func getTableNames() -> [String] {
        // Calls UniFFI-generated function from aliasvault_core.swift
        return getSyncableTableNames()
    }

    /// Merge local and server vaults using LWW strategy.
    ///
    /// - Parameters:
    ///   - localVaultBase64: Base64-encoded encrypted local vault (format: base64(encrypted(base64(sqlite))))
    ///   - serverVaultBase64: Base64-encoded encrypted server vault (format: base64(encrypted(base64(sqlite))))
    ///   - encryptionKey: The encryption key for decrypting vaults
    /// - Returns: Base64-encoded merged vault blob (same format as input)
    public func mergeVaults(
        localVaultBase64: String,
        serverVaultBase64: String,
        encryptionKey: Data
    ) throws -> String {
        // Decrypt both vaults
        // Format is: base64(encrypted(base64(rawSqlite)))
        // Step 1: Decode outer base64 to get encrypted bytes
        guard let localVaultData = Data(base64Encoded: localVaultBase64),
              let serverVaultData = Data(base64Encoded: serverVaultBase64) else {
            throw VaultMergeError.invalidInput("Invalid base64 vault data")
        }

        // Step 2: Decrypt to get inner base64 string
        let localDecryptedBase64 = try decrypt(data: localVaultData, using: encryptionKey)
        let serverDecryptedBase64 = try decrypt(data: serverVaultData, using: encryptionKey)

        // Step 3: Decode inner base64 to get raw SQLite bytes
        guard let localDecryptedString = String(data: localDecryptedBase64, encoding: .utf8),
              let localDecrypted = Data(base64Encoded: localDecryptedString) else {
            throw VaultMergeError.invalidInput("Invalid inner base64 in local vault")
        }
        guard let serverDecryptedString = String(data: serverDecryptedBase64, encoding: .utf8),
              let serverDecrypted = Data(base64Encoded: serverDecryptedString) else {
            throw VaultMergeError.invalidInput("Invalid inner base64 in server vault")
        }

        // Open both SQLite databases
        let localDb = try openDatabase(from: localDecrypted)
        let serverDb = try openDatabase(from: serverDecrypted)
        defer {
            sqlite3_close(localDb)
            sqlite3_close(serverDb)
        }

        // Read all syncable tables from both databases
        let tableNames = getTableNames()
        var localTables: [[String: Any]] = []
        var serverTables: [[String: Any]] = []

        for tableName in tableNames {
            localTables.append([
                "name": tableName,
                "records": try readTable(from: localDb, tableName: tableName)
            ])
            serverTables.append([
                "name": tableName,
                "records": try readTable(from: serverDb, tableName: tableName)
            ])
        }

        // Create merge input and call Rust
        let mergeInput: [String: Any] = [
            "local_tables": localTables,
            "server_tables": serverTables
        ]

        let inputJson = try serializeToJson(mergeInput)
        // Call UniFFI-generated function from aliasvault_core.swift
        let outputJson = try mergeVaultsJson(inputJson: inputJson)
        let output = try parseOutput(outputJson)

        // Apply SQL statements to local database
        for statement in output.statements {
            try executeStatement(on: localDb, sql: statement.sql, params: statement.params)
        }

        // Export, encrypt, and return in same format: base64(encrypted(base64(sqlite)))
        let mergedData = try exportDatabase(from: localDb)
        // Step 1: Base64 encode raw SQLite bytes
        let innerBase64 = mergedData.base64EncodedString()
        // Step 2: Encrypt the base64 string
        let encryptedData = try encrypt(data: Data(innerBase64.utf8), using: encryptionKey)
        // Step 3: Base64 encode the encrypted data
        return encryptedData.base64EncodedString()
    }

    /// Prune expired items from trash.
    ///
    /// - Parameters:
    ///   - vaultBase64: Base64-encoded encrypted vault (format: base64(encrypted(base64(sqlite))))
    ///   - retentionDays: Number of days to keep items in trash (default: 30)
    ///   - encryptionKey: The encryption key for decrypting vault
    /// - Returns: Tuple of (prunedVaultBase64, prunedCount) - same format as input
    public func pruneVault(
        vaultBase64: String,
        retentionDays: Int = 30,
        encryptionKey: Data
    ) throws -> (vaultBase64: String, prunedCount: Int) {
        // Decrypt vault: base64(encrypted(base64(sqlite)))
        // Step 1: Decode outer base64 to get encrypted bytes
        guard let vaultData = Data(base64Encoded: vaultBase64) else {
            throw VaultMergeError.invalidInput("Invalid base64 vault data")
        }

        // Step 2: Decrypt to get inner base64 string
        let decryptedBase64 = try decrypt(data: vaultData, using: encryptionKey)

        // Step 3: Decode inner base64 to get raw SQLite bytes
        guard let decryptedString = String(data: decryptedBase64, encoding: .utf8),
              let decrypted = Data(base64Encoded: decryptedString) else {
            throw VaultMergeError.invalidInput("Invalid inner base64 in vault")
        }

        let db = try openDatabase(from: decrypted)
        defer { sqlite3_close(db) }

        // Read tables needed for pruning
        let pruneTableNames = ["Items", "FieldValues", "Attachments", "TotpCodes", "Passkeys"]
        var tables: [[String: Any]] = []

        for tableName in pruneTableNames {
            tables.append([
                "name": tableName,
                "records": try readTable(from: db, tableName: tableName)
            ])
        }

        // Call Rust prune
        let pruneInput: [String: Any] = [
            "tables": tables,
            "retention_days": retentionDays
        ]

        let inputJson = try serializeToJson(pruneInput)
        // Call UniFFI-generated function from aliasvault_core.swift
        let outputJson = try pruneVaultJson(inputJson: inputJson)
        let output = try parseOutput(outputJson)

        // Apply SQL statements
        for statement in output.statements {
            try executeStatement(on: db, sql: statement.sql, params: statement.params)
        }

        // Export, encrypt, and return in same format: base64(encrypted(base64(sqlite)))
        let prunedData = try exportDatabase(from: db)
        // Step 1: Base64 encode raw SQLite bytes
        let innerBase64 = prunedData.base64EncodedString()
        // Step 2: Encrypt the base64 string
        let encryptedData = try encrypt(data: Data(innerBase64.utf8), using: encryptionKey)
        // Step 3: Base64 encode the encrypted data
        return (encryptedData.base64EncodedString(), output.statements.count)
    }

    // MARK: - Encryption Helpers

    private func encrypt(data: Data, using key: Data) throws -> Data {
        let symmetricKey = SymmetricKey(data: key)
        guard let sealedBox = try? AES.GCM.seal(data, using: symmetricKey),
              let combined = sealedBox.combined else {
            throw VaultMergeError.encryptionError("Encryption failed")
        }
        return combined
    }

    private func decrypt(data: Data, using key: Data) throws -> Data {
        let symmetricKey = SymmetricKey(data: key)
        guard let sealedBox = try? AES.GCM.SealedBox(combined: data),
              let decrypted = try? AES.GCM.open(sealedBox, using: symmetricKey) else {
            throw VaultMergeError.encryptionError("Decryption failed")
        }
        return decrypted
    }

    // MARK: - Database Helpers

    private struct ParsedOutput {
        let statements: [(sql: String, params: [Any])]
    }

    private func openDatabase(from data: Data) throws -> OpaquePointer {
        var db: OpaquePointer?
        guard sqlite3_open(":memory:", &db) == SQLITE_OK else {
            throw VaultMergeError.databaseError("Failed to open in-memory database")
        }

        let count = Int32(data.count)
        let result = data.withUnsafeBytes { (ptr: UnsafeRawBufferPointer) -> Int32 in
            guard let baseAddress = ptr.baseAddress else { return SQLITE_ERROR }
            let pData = sqlite3_malloc(count)
            memcpy(pData, baseAddress, Int(count))
            return sqlite3_deserialize(
                db,
                "main",
                pData?.assumingMemoryBound(to: UInt8.self),
                Int64(count),
                Int64(count),
                UInt32(SQLITE_DESERIALIZE_FREEONCLOSE | SQLITE_DESERIALIZE_RESIZEABLE)
            )
        }

        guard result == SQLITE_OK else {
            sqlite3_close(db)
            throw VaultMergeError.databaseError("Failed to deserialize database: \(result)")
        }

        return db!
    }

    private func exportDatabase(from db: OpaquePointer) throws -> Data {
        var size: sqlite3_int64 = 0
        guard let serialized = sqlite3_serialize(db, "main", &size, 0) else {
            throw VaultMergeError.databaseError("Failed to serialize database")
        }
        let data = Data(bytes: serialized, count: Int(size))
        sqlite3_free(serialized)
        return data
    }

    private func readTable(from db: OpaquePointer, tableName: String) throws -> [[String: Any]] {
        var records: [[String: Any]] = []

        // Check if table exists
        let checkSql = "SELECT name FROM sqlite_master WHERE type='table' AND name=?"
        var checkStmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, checkSql, -1, &checkStmt, nil) == SQLITE_OK else { return records }
        defer { sqlite3_finalize(checkStmt) }

        sqlite3_bind_text(checkStmt, 1, tableName, -1, nil)
        if sqlite3_step(checkStmt) != SQLITE_ROW { return records }

        // Read all records
        let sql = "SELECT * FROM \(tableName)"
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw VaultMergeError.databaseError("Failed to prepare statement for \(tableName)")
        }
        defer { sqlite3_finalize(stmt) }

        let columnCount = sqlite3_column_count(stmt)

        while sqlite3_step(stmt) == SQLITE_ROW {
            var record: [String: Any] = [:]
            for iVal in 0..<columnCount {
                let columnName = String(cString: sqlite3_column_name(stmt, iVal))
                switch sqlite3_column_type(stmt, iVal) {
                case SQLITE_INTEGER:
                    record[columnName] = sqlite3_column_int64(stmt, iVal)
                case SQLITE_FLOAT:
                    record[columnName] = sqlite3_column_double(stmt, iVal)
                case SQLITE_TEXT:
                    if let text = sqlite3_column_text(stmt, iVal) {
                        record[columnName] = String(cString: text)
                    }
                case SQLITE_BLOB:
                    if let blob = sqlite3_column_blob(stmt, iVal) {
                        let size = sqlite3_column_bytes(stmt, iVal)
                        record[columnName] = Data(bytes: blob, count: Int(size)).base64EncodedString()
                    }
                case SQLITE_NULL:
                    record[columnName] = NSNull()
                default:
                    break
                }
            }
            records.append(record)
        }

        return records
    }

    private func executeStatement(on db: OpaquePointer, sql: String, params: [Any]) throws {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            let error = String(cString: sqlite3_errmsg(db))
            throw VaultMergeError.databaseError("Failed to prepare statement: \(error)")
        }
        defer { sqlite3_finalize(stmt) }

        for (index, param) in params.enumerated() {
            let bindIndex = Int32(index + 1)
            switch param {
            case let intValue as Int:
                sqlite3_bind_int64(stmt, bindIndex, Int64(intValue))
            case let int64Value as Int64:
                sqlite3_bind_int64(stmt, bindIndex, int64Value)
            case let doubleValue as Double:
                sqlite3_bind_double(stmt, bindIndex, doubleValue)
            case let stringValue as String:
                sqlite3_bind_text(stmt, bindIndex, stringValue, -1, nil)
            case is NSNull:
                sqlite3_bind_null(stmt, bindIndex)
            default:
                sqlite3_bind_text(stmt, bindIndex, "\(param)", -1, nil)
            }
        }

        guard sqlite3_step(stmt) == SQLITE_DONE else {
            let error = String(cString: sqlite3_errmsg(db))
            throw VaultMergeError.databaseError("Failed to execute statement: \(error)")
        }
    }

    private func serializeToJson(_ dictionary: [String: Any]) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: dictionary, options: [])
        guard let json = String(data: data, encoding: .utf8) else {
            throw VaultMergeError.jsonError("Failed to convert to UTF-8 string")
        }
        return json
    }

    private func parseOutput(_ json: String) throws -> ParsedOutput {
        guard let data = json.data(using: .utf8),
              let rawOutput = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let statementsArray = rawOutput["statements"] as? [[String: Any]] else {
            throw VaultMergeError.jsonError("Failed to parse output JSON")
        }

        let statements = try statementsArray.map { stmt -> (sql: String, params: [Any]) in
            guard let sql = stmt["sql"] as? String,
                  let params = stmt["params"] as? [Any] else {
                throw VaultMergeError.jsonError("Invalid statement format")
            }
            return (sql: sql, params: params)
        }

        return ParsedOutput(statements: statements)
    }
}

File: apps/mobile-app/android/app/src/main/java/net/aliasvault/app/rustcore/VaultMergeService.kt
package net.aliasvault.app.rustcore

import android.database.sqlite.SQLiteDatabase
import android.util.Base64
import android.util.Log
import net.aliasvault.app.vaultstore.VaultCrypto
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Service for vault merge operations using the Rust core library.
 * Wraps UniFFI-generated bindings for LWW merge operations on SQLite vault databases.
 */
object VaultMergeService {
    private const val TAG = "VaultMergeService"

    init {
        // Ensure JNA is initialized before loading the Rust library
        JnaInitializer.ensureInitialized()
        System.loadLibrary("aliasvault_core")
        Log.d(TAG, "Rust core library loaded")
    }

    /**
     * Base exception for vault merge operations.
     */
    sealed class VaultMergeException(message: String) : Exception(message) {
        /**
         * Exception thrown when JSON parsing or serialization fails.
         */
        class JsonError(message: String) : VaultMergeException("JSON error: $message")

        /**
         * Exception thrown when Rust core operations fail.
         */
        class RustError(message: String) : VaultMergeException("Rust error: $message")

        /**
         * Exception thrown when database operations fail.
         */
        class DatabaseError(message: String) : VaultMergeException("Database error: $message")

        /**
         * Exception thrown when input validation fails.
         */
        class InvalidInput(message: String) : VaultMergeException("Invalid input: $message")
    }

    /**
     * Get the list of syncable table names from Rust.
     */
    fun getTableNames(): List<String> {
        return uniffi.aliasvault_core.getSyncableTableNames()
    }

    /**
     * Merge local and server vaults using LWW strategy.
     */
    @Suppress("SwallowedException") // Exceptions are logged and re-thrown with context
    @Throws(VaultMergeException::class)
    fun mergeVaults(
        localVaultBase64: String,
        serverVaultBase64: String,
        encryptionKey: ByteArray,
        tempDir: File,
    ): String {
        // Decrypt both vaults
        // Vault format: base64(encrypted(base64(sqlite_bytes)))
        // After AES decrypt we get base64(sqlite_bytes), need to decode again
        val localDecrypted = try {
            val decryptedBase64 = VaultCrypto.decrypt(Base64.decode(localVaultBase64, Base64.DEFAULT), encryptionKey)
            Base64.decode(decryptedBase64, Base64.NO_WRAP)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to decrypt local vault", e)
            throw VaultMergeException.InvalidInput("Failed to decrypt local vault: ${e.message}")
        }

        val serverDecrypted = try {
            val decryptedBase64 = VaultCrypto.decrypt(Base64.decode(serverVaultBase64, Base64.DEFAULT), encryptionKey)
            Base64.decode(decryptedBase64, Base64.NO_WRAP)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to decrypt server vault", e)
            throw VaultMergeException.InvalidInput("Failed to decrypt server vault: ${e.message}")
        }

        // Create temporary database files
        // Ensure the temp directory exists (may not on first app run or after cache clear)
        if (!tempDir.exists()) {
            tempDir.mkdirs()
        }

        // Verify directory exists and is writable
        if (!tempDir.exists() || !tempDir.isDirectory) {
            throw VaultMergeException.DatabaseError("Temp directory does not exist or is not a directory: ${tempDir.absolutePath}")
        }
        if (!tempDir.canWrite()) {
            throw VaultMergeException.DatabaseError("Temp directory is not writable: ${tempDir.absolutePath}")
        }

        val localDbFile = File(tempDir, "local_merge_temp.db")
        val serverDbFile = File(tempDir, "server_merge_temp.db")

        try {
            // Write decrypted databases to temp files using FileOutputStream to ensure proper flushing
            java.io.FileOutputStream(localDbFile).use { fos ->
                fos.write(localDecrypted)
                fos.fd.sync()
            }

            java.io.FileOutputStream(serverDbFile).use { fos ->
                fos.write(serverDecrypted)
                fos.fd.sync()
            }

            // Verify files were written
            if (!localDbFile.exists() || localDbFile.length() == 0L) {
                throw VaultMergeException.DatabaseError(
                    "Failed to write local temp db: exists=${localDbFile.exists()}, len=${localDbFile.length()}",
                )
            }
            if (!serverDbFile.exists() || serverDbFile.length() == 0L) {
                throw VaultMergeException.DatabaseError(
                    "Failed to write server temp db: exists=${serverDbFile.exists()}, len=${serverDbFile.length()}",
                )
            }

            val localDb = SQLiteDatabase.openDatabase(localDbFile.absolutePath, null, SQLiteDatabase.OPEN_READWRITE)
            val serverDb = SQLiteDatabase.openDatabase(serverDbFile.absolutePath, null, SQLiteDatabase.OPEN_READONLY)

            try {
                // Read all syncable tables
                val tableNames = getTableNames()
                val localTables = JSONArray()
                val serverTables = JSONArray()

                for (tableName in tableNames) {
                    localTables.put(
                        JSONObject().apply {
                            put("name", tableName)
                            put("records", readTable(localDb, tableName))
                        },
                    )
                    serverTables.put(
                        JSONObject().apply {
                            put("name", tableName)
                            put("records", readTable(serverDb, tableName))
                        },
                    )
                }

                // Call Rust merge
                val mergeInput = JSONObject().apply {
                    put("local_tables", localTables)
                    put("server_tables", serverTables)
                }

                val outputJson = uniffi.aliasvault_core.mergeVaultsJson(mergeInput.toString())
                val statements = parseStatements(outputJson)

                // Apply SQL statements
                localDb.beginTransaction()
                try {
                    for ((sql, params) in statements) {
                        localDb.execSQL(sql, params.toTypedArray())
                    }
                    localDb.setTransactionSuccessful()
                } finally {
                    localDb.endTransaction()
                }

                localDb.close()
                serverDb.close()

                // Encrypt and return
                // Output format must match input: base64(encrypted(base64(sqlite_bytes)))
                val mergedData = localDbFile.readBytes()
                val mergedBase64 = Base64.encodeToString(mergedData, Base64.NO_WRAP)
                val encrypted = VaultCrypto.encrypt(mergedBase64.toByteArray(Charsets.UTF_8), encryptionKey)
                return Base64.encodeToString(encrypted, Base64.DEFAULT)
            } finally {
                if (localDb.isOpen) localDb.close()
                if (serverDb.isOpen) serverDb.close()
            }
        } finally {
            localDbFile.delete()
            serverDbFile.delete()
        }
    }

    /**
     * Prune expired items from trash.
     */
    @Suppress("SwallowedException") // Exceptions are logged and re-thrown with context
    @Throws(VaultMergeException::class)
    fun pruneVault(
        vaultBase64: String,
        retentionDays: Int = 30,
        encryptionKey: ByteArray,
        tempDir: File,
    ): Pair<String, Int> {
        // Vault format: base64(encrypted(base64(sqlite_bytes)))
        // After AES decrypt we get base64(sqlite_bytes), need to decode again
        val decrypted = try {
            val decryptedBase64 = VaultCrypto.decrypt(Base64.decode(vaultBase64, Base64.DEFAULT), encryptionKey)
            Base64.decode(decryptedBase64, Base64.NO_WRAP)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to decrypt vault for pruning", e)
            throw VaultMergeException.InvalidInput("Failed to decrypt vault: ${e.message}")
        }

        // Ensure the temp directory exists (may not on first app run or after cache clear)
        if (!tempDir.exists()) {
            tempDir.mkdirs()
        }
        val dbFile = File(tempDir, "prune_temp.db")

        try {
            dbFile.writeBytes(decrypted)
            val db = SQLiteDatabase.openDatabase(dbFile.absolutePath, null, SQLiteDatabase.OPEN_READWRITE)

            try {
                // Read tables needed for pruning
                val pruneTableNames = listOf("Items", "FieldValues", "Attachments", "TotpCodes", "Passkeys")
                val tables = JSONArray()

                for (tableName in pruneTableNames) {
                    tables.put(
                        JSONObject().apply {
                            put("name", tableName)
                            put("records", readTable(db, tableName))
                        },
                    )
                }

                // Call Rust prune - use ISO8601 format: YYYY-MM-DDTHH:MM:SS.sssZ
                val dateFormat = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", java.util.Locale.US)
                dateFormat.timeZone = java.util.TimeZone.getTimeZone("UTC")
                val currentTime = dateFormat.format(java.util.Date())

                val pruneInput = JSONObject().apply {
                    put("tables", tables)
                    put("retention_days", retentionDays)
                    put("current_time", currentTime)
                }

                val outputJson = uniffi.aliasvault_core.pruneVaultJson(pruneInput.toString())
                val statements = parseStatements(outputJson)

                // Apply SQL statements
                db.beginTransaction()
                try {
                    for ((sql, params) in statements) {
                        db.execSQL(sql, params.toTypedArray())
                    }
                    db.setTransactionSuccessful()
                } finally {
                    db.endTransaction()
                }

                db.close()

                // Encrypt and return
                // Output format must match input: base64(encrypted(base64(sqlite_bytes)))
                val prunedData = dbFile.readBytes()
                val prunedBase64 = Base64.encodeToString(prunedData, Base64.NO_WRAP)
                val encrypted = VaultCrypto.encrypt(prunedBase64.toByteArray(Charsets.UTF_8), encryptionKey)
                return Pair(Base64.encodeToString(encrypted, Base64.DEFAULT), statements.size)
            } finally {
                if (db.isOpen) db.close()
            }
        } finally {
            dbFile.delete()
        }
    }

    private fun readTable(db: SQLiteDatabase, tableName: String): JSONArray {
        val records = JSONArray()

        // Check if table exists
        val checkCursor = db.rawQuery(
            "SELECT name FROM sqlite_master WHERE type='table' AND name=?",
            arrayOf(tableName),
        )
        val tableExists = checkCursor.moveToFirst()
        checkCursor.close()
        if (!tableExists) return records

        // Read all records
        val cursor = db.rawQuery("SELECT * FROM $tableName", null)
        try {
            while (cursor.moveToNext()) {
                val record = JSONObject()
                for (i in 0 until cursor.columnCount) {
                    val columnName = cursor.getColumnName(i)
                    val value = when (cursor.getType(i)) {
                        android.database.Cursor.FIELD_TYPE_INTEGER -> cursor.getLong(i)
                        android.database.Cursor.FIELD_TYPE_FLOAT -> cursor.getDouble(i)
                        android.database.Cursor.FIELD_TYPE_STRING -> cursor.getString(i)
                        android.database.Cursor.FIELD_TYPE_BLOB -> Base64.encodeToString(cursor.getBlob(i), Base64.DEFAULT)
                        else -> JSONObject.NULL
                    }
                    record.put(columnName, value)
                }
                records.put(record)
            }
        } finally {
            cursor.close()
        }

        return records
    }

    private fun parseStatements(json: String): List<Pair<String, List<String?>>> {
        val output = JSONObject(json)
        val statementsArray = output.optJSONArray("statements") ?: JSONArray()
        val statements = mutableListOf<Pair<String, List<String?>>>()

        for (i in 0 until statementsArray.length()) {
            val stmt = statementsArray.getJSONObject(i)
            val sql = stmt.getString("sql")
            val paramsArray = stmt.optJSONArray("params") ?: JSONArray()
            val params = (0 until paramsArray.length()).map { j ->
                val param = paramsArray.opt(j)
                if (param == JSONObject.NULL || param == null) null else param.toString()
            }
            statements.add(Pair(sql, params))
        }

        return statements
    }
}
